// react-scroll-list/lib/ScrollListItem.js
'use strict';Object.defineProperty(exports, "__esModule", { value: true });var _createClass = function () {function defineProperties(target, props) {for (var i = 0; i < props.length; i++) {var descriptor = props[i];descriptor.enumerable = descriptor.enumerable || false;descriptor.configurable = true;if ("value" in descriptor) descriptor.writable = true;Object.defineProperty(target, descriptor.key, descriptor);}}return function (Constructor, protoProps, staticProps) {if (protoProps) defineProperties(Constructor.prototype, protoProps);if (staticProps) defineProperties(Constructor, staticProps);return Constructor;};}();var _react = require('react');var _react2 = _interopRequireDefault(_react);
var _propTypes = require('prop-types');var _propTypes2 = _interopRequireDefault(_propTypes);
var _raf = require('raf');var _raf2 = _interopRequireDefault(_raf);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}function _classCallCheck(instance, Constructor) {if (!(instance instanceof Constructor)) {throw new TypeError("Cannot call a class as a function");}}function _possibleConstructorReturn(self, call) {if (!self) {throw new ReferenceError("this hasn't been initialised - super() hasn't been called");}return call && (typeof call === "object" || typeof call === "function") ? call : self;}function _inherits(subClass, superClass) {if (typeof superClass !== "function" && superClass !== null) {throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);}subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } });if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;}var

ScrollListItem = function (_Component) {_inherits(ScrollListItem, _Component);function ScrollListItem() {var _ref;var _temp, _this, _ret;_classCallCheck(this, ScrollListItem);for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {args[_key] = arguments[_key];}return _ret = (_temp = (_this = _possibleConstructorReturn(this, (_ref = ScrollListItem.__proto__ || Object.getPrototypeOf(ScrollListItem)).call.apply(_ref, [this].concat(args))), _this), _this.















    unmounting = false, _this.
    rafId = null, _this.
    lastHeight = null, _this.

    measureHeight = function () {
      if (_this.unmounting || !_this.element) {
        return;
      }var _this$props =

      _this.props,index = _this$props.index,onRectChange = _this$props.onRectChange;
      var rect = _this.element.getBoundingClientRect();

      if (_this.lastHeight !== rect.height) {
        onRectChange(index, rect);
      }

      _this.lastHeight = rect.height;

      _this.rafId = (0, _raf2.default)(_this.measureHeight);
    }, _temp), _possibleConstructorReturn(_this, _ret);}_createClass(ScrollListItem, [{ key: 'componentDidMount', value: function componentDidMount() {this.measureHeight();} }, { key: 'componentWillUnmount', value: function componentWillUnmount() {this.unmounting = true;_raf2.default.cancel(this.rafId);} }, { key: 'render', value: function render()

    {var _this2 = this;
      /* eslint-disable react/prop-types */var
      children = this.props.children;
      /* eslint-enable react/prop-types */

      return (
        _react2.default.createElement('div', {
            ref: function ref(element) {
              _this2.element = element;
            } },

          children));


    } }]);return ScrollListItem;}(_react.Component);ScrollListItem.propTypes = { index: _propTypes2.default.number.isRequired, onRectChange: _propTypes2.default.func.isRequired };exports.default =


ScrollListItem;
